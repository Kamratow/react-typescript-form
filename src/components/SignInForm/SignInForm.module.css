.pageWrapper {
    width: 30%;
    margin: auto;
}
