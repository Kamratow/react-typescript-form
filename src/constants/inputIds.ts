export default class inputIds {
    static 'EMAIL' = 'emailInput';
    static 'PASSWORD' = 'passwordInput';
    static 'FULL_NAME' = 'fullNameInput';
    static 'CONFIRM_PASSWORD' = 'confirmPasswordInput';
}