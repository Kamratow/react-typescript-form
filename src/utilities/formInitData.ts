import { IFormInput } from "../models/formInput.model";
import inputIds from "../constants/inputIds";
import textConstants from "../constants/textConstants";

export const getInitialSignInFormData = (): IFormInput[] => {
    return [
        {
            id: inputIds.EMAIL,
            value: "",
            type: "email",
            label: textConstants.EMAIL_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.PASSWORD,
            value: "",
            type: "password",
            label: textConstants.PASSWORD_INPUT_LABEL,
            errorMessage: "",
        },
    ];
};

export const getInitialSignUpFormData = (): IFormInput[] => {
    return [
        {
            id: inputIds.EMAIL,
            value: "",
            type: "email",
            label: textConstants.EMAIL_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.FULL_NAME,
            value: "",
            type: "text",
            label: textConstants.FULL_NAME_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.PASSWORD,
            value: "",
            type: "password",
            label: textConstants.PASSWORD_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.CONFIRM_PASSWORD,
            value: "",
            type: "password",
            label: textConstants.CONFIRM_PASSWORD_INPUT_LABEL,
            errorMessage: "",
        }
    ];
};