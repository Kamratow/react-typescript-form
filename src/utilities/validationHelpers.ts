import {IFormInput} from "../models/formInput.model";
import inputIds from "../constants/inputIds";

export const validateAndParseInputData = (newValue: string, formInput: IFormInput) => {
    let parsedAndValidatedInput: IFormInput = {
        ...formInput,
        value: newValue,
        errorMessage: ''
    };

    //validate if input is empty
    if (newValue === '') {
        if (parsedAndValidatedInput.id === inputIds.CONFIRM_PASSWORD) {
            parsedAndValidatedInput.errorMessage = "Please confirm your password";
        } else {
            parsedAndValidatedInput.errorMessage = `${parsedAndValidatedInput.label} is required`;
        }

        return parsedAndValidatedInput;
    }

    //validate if value matches input type
    switch (formInput.type) {
        case 'email':
            parsedAndValidatedInput = validateEmailValue(parsedAndValidatedInput);
            break;
        case 'password':
            parsedAndValidatedInput = validatePasswordValue(parsedAndValidatedInput);
            break;
        default:
            break;
    }

    //check if input is full name
    if (parsedAndValidatedInput.id === inputIds.FULL_NAME) {
        parsedAndValidatedInput = validateFullNameValue(parsedAndValidatedInput);
    }

    return parsedAndValidatedInput;
};

export const validateEmailValue = (formInput: IFormInput): IFormInput => {
    const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    if (!emailRegex.test(formInput.value)) {
        formInput.errorMessage = 'Invalid e-mail address';
    }

    return formInput;
};

export const validatePasswordValue = (formInput: IFormInput): IFormInput => {
    if (formInput.value.length < 8) {
        formInput.errorMessage = 'Password is too short';
    }

    return formInput;
};

export const validateFullNameValue = (formInput: IFormInput): IFormInput => {
    const fullNameRegex = /([a-zA-Z]+\s+[a-zA-Z]+)+/;

    if (!fullNameRegex.test(formInput.value)) {
        formInput.errorMessage = 'Missing last name';
    }

    return formInput;
};

export const checkIsFormDataValid = (formData: IFormInput[]): boolean => {
    return formData.filter(singleInput => singleInput.value !== '').length === formData.length && !formData.find(singleInput => singleInput.errorMessage !== '');
};

export const validateConfirmPasswordValue = (formData: IFormInput[]): IFormInput[] => {
    const validatedFormData = [...formData];
    const passwordInput = validatedFormData.find(singleInput => singleInput.id === inputIds.PASSWORD);
    const confirmPasswordInput = validatedFormData.find(singleInput => singleInput.id === inputIds.CONFIRM_PASSWORD);

    if (passwordInput && confirmPasswordInput && passwordInput.value !== '' && confirmPasswordInput.value !== '' && passwordInput.value !== confirmPasswordInput.value) {
        validatedFormData[validatedFormData.findIndex(singleInput => singleInput.id === inputIds.CONFIRM_PASSWORD)].errorMessage = 'Please confirm your password';
    }

    return validatedFormData;
};
