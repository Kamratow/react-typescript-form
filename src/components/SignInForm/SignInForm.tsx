import React, { useState } from "react";

import Header from "../common/Header/Header";
import Form from "../common/Form/Form";
import StandardField from "../common/StandardField/StandardField";
import FormLink from "../common/FormLink/FormLink";
import SubmitButton from "../common/SubmitButton/SubmitButton";

import { getInitialSignInFormData } from "../../utilities/formInitData";
import {
    validateAndParseInputData,
    checkIsFormDataValid,
} from "../../utilities/validationHelpers";

import textConstants from "../../constants/textConstants";

import { IFormInput } from "../../models/formInput.model";

const SignInForm: React.FunctionComponent = () => {
    const [formData, setFormData] = useState<IFormInput[]>(
        getInitialSignInFormData()
    );

    const handleInputChange = (
        newValue: string,
        formInputData: IFormInput,
        inputIndex: number
    ): void => {
        const newFormData = [...formData];
        const newFormInputData = validateAndParseInputData(
            newValue,
            formInputData
        );
        newFormData[inputIndex] = {
            ...newFormInputData,
        };
        setFormData(newFormData);
    };

    const getFormInputsToRender = () => {
        return formData.map((singleInput, index) => {
            return (
                <StandardField
                    key={singleInput.id}
                    id={singleInput.id}
                    label={singleInput.label}
                    type={singleInput.type}
                    value={singleInput.value}
                    changeHandler={(newInputValue: string) =>
                        handleInputChange(newInputValue, singleInput, index)
                    }
                    errorMessage={singleInput.errorMessage}
                />
            );
        });
    };

    return (
        <>
            <Header headerContent={textConstants.SIGNIN_HEADER} />
            <Form>
                {getFormInputsToRender()}
                <SubmitButton
                    disabled={!checkIsFormDataValid(formData)}
                    buttonLabel={textConstants.FORM_SUBMIT_BUTTON}
                />
            </Form>
            <FormLink
                description={textConstants.SIGNIN_FORMLINK_DESCRIPTION}
                linkPath={textConstants.SIGNIN_FORMLINK_PATH}
                linkText={textConstants.SIGNIN_FORMLINK_LINK_CONTENT}
            />
        </>
    );
};

export default SignInForm;
