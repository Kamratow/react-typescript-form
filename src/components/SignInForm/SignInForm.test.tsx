import React from "react";
import { shallow } from "enzyme";

import SignInForm from "./SignInForm";
import inputIds from "../../constants/inputIds";

import * as formInitData from "../../utilities/formInitData";

import {
    validSignInFormData,
    invalidSignInFormData,
} from "./testMocks/signInFormDataMocks";

describe("SignInForm", () => {
    it("should render with proper input configuration", () => {
        const wrapper = shallow(<SignInForm />);

        expect(wrapper).toMatchSnapshot();
    });

    describe("EmailInput", () => {
        let wrapper: any;
        let emailInput: any;

        beforeEach(() => {
            wrapper = shallow(<SignInForm />);
            emailInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.EMAIL
                )
                .at(0);
        });

        it("should be properly validated when email is incorrect", () => {
            const expectedErrorMessage = "Invalid e-mail address";
            const invalidEmail = "email";

            emailInput.prop("changeHandler")(invalidEmail);

            emailInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.EMAIL
                )
                .at(0);

            expect(emailInput.prop("value")).toBe(invalidEmail);
            expect(emailInput.prop("errorMessage")).toBe(expectedErrorMessage);
        });

        it("should be properly validated when email is correct", () => {
            const validEmail = "test@test.com";

            emailInput.prop("changeHandler")(validEmail);

            emailInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.EMAIL
                )
                .at(0);

            expect(emailInput.prop("value")).toBe(validEmail);
            expect(emailInput.prop("errorMessage")).toBe("");
        });

        it("should be properly validated when email is empty", () => {
            const expectedErrorMessage = "E-mail address is required";
            const invalidEmail = "";

            emailInput.prop("changeHandler")(invalidEmail);

            emailInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.EMAIL
                )
                .at(0);

            expect(emailInput.prop("value")).toBe(invalidEmail);
            expect(emailInput.prop("errorMessage")).toBe(expectedErrorMessage);
        });
    });

    describe("PasswordInput", () => {
        let wrapper: any;
        let passwordInput: any;

        beforeEach(() => {
            wrapper = shallow(<SignInForm />);
            passwordInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.PASSWORD
                )
                .at(0);
        });

        it("should be validated properly when password is incorrect", () => {
            const expectedErrorMessage = "Password is too short";
            const invalidPassword = "1234";

            passwordInput.prop("changeHandler")(invalidPassword);

            passwordInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.PASSWORD
                )
                .at(0);

            expect(passwordInput.prop("value")).toBe(invalidPassword);
            expect(passwordInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });

        it("should be validated properly when password is correct", () => {
            const validPassword = "12345678";

            passwordInput.prop("changeHandler")(validPassword);

            passwordInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.PASSWORD
                )
                .at(0);

            expect(passwordInput.prop("value")).toBe(validPassword);
            expect(passwordInput.prop("errorMessage")).toBe("");
        });

        it("should be validated properly when password is empty", () => {
            const expectedErrorMessage = "Password is required";
            const invalidPassword = "";

            passwordInput.prop("changeHandler")(invalidPassword);

            passwordInput = wrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.PASSWORD
                )
                .at(0);

            expect(passwordInput.prop("value")).toBe(invalidPassword);
            expect(passwordInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });
    });

    describe("SubmitButton", () => {
        it("should be disabled if form data is invalid", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignInFormData"
            ).mockImplementation(() => invalidSignInFormData);

            const wrapper = shallow(<SignInForm />);

            expect(wrapper.find("SubmitButton").prop("disabled")).toBe(true);
        });

        it("should not be disabled if form data is valid", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignInFormData"
            ).mockImplementation(() => validSignInFormData);

            const wrapper = shallow(<SignInForm />);

            expect(wrapper.find("SubmitButton").prop("disabled")).toBe(false);
        });
    });
});
