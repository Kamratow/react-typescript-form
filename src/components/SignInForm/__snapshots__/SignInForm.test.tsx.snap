// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm should render with proper input configuration 1`] = `
<Fragment>
  <Header
    headerContent="Sign in"
  />
  <Form>
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="emailInput"
      key="emailInput"
      label="E-mail address"
      type="email"
      value=""
    />
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="passwordInput"
      key="passwordInput"
      label="Password"
      type="password"
      value=""
    />
    <SubmitButton
      buttonLabel="Continue"
      disabled={true}
    />
  </Form>
  <FormLink
    description="Not our member yet?"
    linkPath="/signup"
    linkText="Click here to create new account"
  />
</Fragment>
`;
